File: sandbox/telemetry.py
"""OpenTelemetry runtime for the Sandbox HTTP process.

The durable MySQL trace projection remains the product query source. This
module is the runtime exporter/propagator used for distributed tracing; it
never records request bodies, tool arguments, credentials, or response data.
"""

from __future__ import annotations

import os
from contextlib import contextmanager
from typing import Iterator

from opentelemetry import propagate, trace
from opentelemetry.context import Context
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.trace import (
    NonRecordingSpan,
    Span,
    SpanContext,
    SpanKind,
    Status,
    StatusCode,
    TraceFlags,
    TraceState,
)
from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator

_provider: TracerProvider | None = None
_shutdown = False


def _positive_int(value: object, fallback: int) -> int:
    try:
        parsed = int(str(value).strip())
    except (TypeError, ValueError):
        return fallback
    return parsed if parsed > 0 else fallback


def _endpoint() -> str | None:
    traces = os.getenv("OTEL_EXPORTER_OTLP_TRACES_ENDPOINT", "").strip()
    if traces:
        return traces
    base = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "").strip().rstrip("/")
    return f"{base}/v1/traces" if base else None


def start_telemetry(service_name: str = "pi-enterprise-sandbox") -> TracerProvider:
    """Install a process-wide provider and W3C propagator once."""

    global _provider, _shutdown
    if _provider is not None:
        return _provider

    # Explicitly select W3C even when no exporter is configured. Propagation
    # and local spans are useful in development and do not require a backend.
    propagate.set_global_textmap(TraceContextTextMapPropagator())
    provider = TracerProvider(
        resource=Resource.create(
            {
                "service.name": os.getenv("OTEL_SERVICE_NAME", service_name),
                "service.version": "4.0.0",
                "deployment.environment.name": os.getenv(
                    "DEPLOYMENT_ENV", os.getenv("SANDBOX_DEPLOYMENT_ENV", "development")
                ),
            }
        )
    )
    endpoint = _endpoint()
    if endpoint and os.getenv("OTEL_SDK_DISABLED", "").strip().lower() != "true":
        exporter = OTLPSpanExporter(
            endpoint=endpoint,
            timeout=_positive_int(os.getenv("OTEL_EXPORTER_OTLP_TIMEOUT"), 10),
        )
        provider.add_span_processor(
            BatchSpanProcessor(
                exporter,
                max_queue_size=_positive_int(os.getenv("OTEL_BSP_MAX_QUEUE_SIZE"), 2048),
                max_export_batch_size=_positive_int(
                    os.getenv("OTEL_BSP_MAX_EXPORT_BATCH_SIZE"), 512
                ),
                schedule_delay_millis=_positive_int(
                    os.getenv("OTEL_BSP_SCHEDULE_DELAY"), 5000
                ),
                export_timeout_millis=_positive_int(
                    os.getenv("OTEL_BSP_EXPORT_TIMEOUT"), 10000
                ),
            )
        )
    trace.set_tracer_provider(provider)
    _provider = provider
    _shutdown = False
    return provider


def shutdown_telemetry() -> None:
    global _shutdown
    if _shutdown or _provider is None:
        return
    _shutdown = True
    _provider.shutdown()


def tracer(service_name: str = "pi-enterprise-sandbox"):
    start_telemetry(service_name)
    return trace.get_tracer(service_name, "4.0.0")


def extracted_parent(headers: object) -> Context:
    """Extract a W3C parent from a Starlette headers mapping."""

    try:
        return propagate.extract(dict(headers))
    except (TypeError, ValueError):
        return Context()


def synthetic_trace_parent(trace_id: str, span_id: str) -> Context | None:
    """Build a remote parent for legacy X-Trace-Id-only callers."""

    if (
        not isinstance(trace_id, str)
        or not isinstance(span_id, str)
        or len(trace_id) != 32
        or len(span_id) != 16
        or trace_id == "0" * 32
        or span_id == "0" * 16
    ):
        return None
    try:
        parent = SpanContext(
            trace_id=int(trace_id, 16),
            span_id=int(span_id, 16),
            is_remote=True,
            trace_flags=TraceFlags(TraceFlags.SAMPLED),
            trace_state=TraceState(),
        )
    except (TypeError, ValueError):
        return None
    return trace.set_span_in_context(NonRecordingSpan(parent))


@contextmanager
def server_span(
    method: str,
    path: str,
    parent: Context | None = None,
) -> Iterator[Span]:
    """Yield a server span with only low-cardinality safe attributes."""

    span = tracer().start_span(
        f"{method} {path}",
        context=parent,
        kind=SpanKind.SERVER,
        attributes={
            "http.request.method": method,
            "url.path": path,
        },
    )
    with trace.use_span(span, end_on_exit=True):
        try:
            yield span
        except Exception as exc:
            span.record_exception(exc)
            span.set_status(Status(StatusCode.ERROR, str(exc)))
            raise


def finish_server_span(span: Span, status_code: int) -> None:
    span.set_attribute("http.response.status_code", int(status_code))
    if status_code >= 500:
        span.set_status(Status(StatusCode.ERROR))
    elif status_code >= 100:
        span.set_status(Status(StatusCode.OK))

File: api-server/application/telemetry.js
import {
  ROOT_CONTEXT,
  SpanKind,
  SpanStatusCode,
  TraceFlags,
  createTraceState,
  context,
  propagation,
  trace,
} from '@opentelemetry/api';
import { W3CTraceContextPropagator } from '@opentelemetry/core';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-proto';
import { resourceFromAttributes } from '@opentelemetry/resources';
import { NodeSDK } from '@opentelemetry/sdk-node';
import {
  BatchSpanProcessor,
  NoopSpanProcessor,
} from '@opentelemetry/sdk-trace-base';
import { UndiciInstrumentation } from '@opentelemetry/instrumentation-undici';
import { ATTR_SERVICE_NAME, ATTR_SERVICE_VERSION } from '@opentelemetry/semantic-conventions';

const INSTRUMENTATION_NAME = 'pi-enterprise-bff';
const DEFAULT_EXPORT_TIMEOUT_MS = 10_000;
const DEFAULT_SCHEDULE_DELAY_MS = 5_000;
const DEFAULT_MAX_QUEUE_SIZE = 2_048;
const DEFAULT_MAX_EXPORT_BATCH_SIZE = 512;

let telemetry = null;

function positiveInteger(value, fallback) {
  const parsed = Number(value);
  return Number.isSafeInteger(parsed) && parsed > 0 ? parsed : fallback;
}

function configuredEndpoint(env) {
  const traces = String(env.OTEL_EXPORTER_OTLP_TRACES_ENDPOINT || '').trim();
  if (traces) return traces;
  const base = String(env.OTEL_EXPORTER_OTLP_ENDPOINT || '').trim().replace(/\/$/, '');
  return base ? `${base}/v1/traces` : null;
}

export async function startTelemetry(env = process.env) {
  if (telemetry) return telemetry;
  const endpoint = configuredEndpoint(env);
  propagation.setGlobalPropagator(new W3CTraceContextPropagator());
  if (String(env.OTEL_SDK_DISABLED || '').toLowerCase() === 'true') {
    telemetry = Object.freeze({ enabled: false, async shutdown() {} });
    return telemetry;
  }

  const spanProcessor = endpoint
    ? new BatchSpanProcessor(
        new OTLPTraceExporter({
          url: endpoint,
          timeoutMillis: positiveInteger(
            env.OTEL_EXPORTER_OTLP_TIMEOUT,
            DEFAULT_EXPORT_TIMEOUT_MS,
          ),
        }),
        {
          maxQueueSize: positiveInteger(
            env.OTEL_BSP_MAX_QUEUE_SIZE,
            DEFAULT_MAX_QUEUE_SIZE,
          ),
          maxExportBatchSize: positiveInteger(
            env.OTEL_BSP_MAX_EXPORT_BATCH_SIZE,
            DEFAULT_MAX_EXPORT_BATCH_SIZE,
          ),
          scheduledDelayMillis: positiveInteger(
            env.OTEL_BSP_SCHEDULE_DELAY,
            DEFAULT_SCHEDULE_DELAY_MS,
          ),
          exportTimeoutMillis: positiveInteger(
            env.OTEL_BSP_EXPORT_TIMEOUT,
            DEFAULT_EXPORT_TIMEOUT_MS,
          ),
        },
      )
    : new NoopSpanProcessor();
  const sdk = new NodeSDK({
    resource: resourceFromAttributes({
      [ATTR_SERVICE_NAME]: String(env.OTEL_SERVICE_NAME || 'pi-enterprise-bff'),
      [ATTR_SERVICE_VERSION]: '4.0.0',
      'deployment.environment.name': String(
        env.DEPLOYMENT_ENV || env.NODE_ENV || 'development',
      ),
    }),
    spanProcessors: [spanProcessor],
    textMapPropagator: new W3CTraceContextPropagator(),
    instrumentations: [
      new UndiciInstrumentation({
        requireParentforSpans: true,
        headersToSpanAttributes: {
          requestHeaders: [],
          responseHeaders: [],
        },
      }),
    ],
  });
  sdk.start();
  let stopped = false;
  telemetry = Object.freeze({
    enabled: Boolean(endpoint),
    endpoint,
    async shutdown() {
      if (stopped) return;
      stopped = true;
      await sdk.shutdown();
    },
  });
  return telemetry;
}

export function requestCarrier(headers = {}) {
  const carrier = {};
  for (const name of ['traceparent', 'tracestate']) {
    const value = headers[name] ?? headers[name.toLowerCase()];
    if (typeof value === 'string' && value) carrier[name] = value;
  }
  return carrier;
}

export function contextFromTraceContext(value) {
  const traceId = String(value?.traceId || '').toLowerCase();
  const spanId = String(value?.parentSpanId || value?.spanId || '').toLowerCase();
  if (!/^[0-9a-f]{32}$/.test(traceId) || !/^[0-9a-f]{16}$/.test(spanId)) {
    return ROOT_CONTEXT;
  }
  const flags = Number.parseInt(String(value?.traceFlags || '01'), 16);
  return trace.setSpanContext(ROOT_CONTEXT, {
    traceId,
    spanId,
    traceFlags: Number.isFinite(flags) ? flags : TraceFlags.SAMPLED,
    isRemote: true,
    ...(value?.tracestate ? { traceState: createTraceState(value.tracestate) } : {}),
  });
}

export function startHttpServerSpan(req, resolvedTraceContext) {
  const parent = propagation.extract(ROOT_CONTEXT, requestCarrier(req?.headers));
  const parentContext = trace.getSpanContext(parent)
    ? parent
    : contextFromTraceContext(resolvedTraceContext);
  const tracer = trace.getTracer(INSTRUMENTATION_NAME, '4.0.0');
  const span = tracer.startSpan(
    `${String(req?.method || 'HTTP')} ${String(req?.url || '/').split('?')[0]}`,
    {
      kind: SpanKind.SERVER,
      attributes: {
        'http.request.method': String(req?.method || 'UNKNOWN'),
        'url.path': String(req?.url || '/').split('?')[0],
      },
    },
    parentContext,
  );
  const activeContext = trace.setSpan(parentContext, span);
  let ended = false;
  const end = (error, statusCodeValue) => {
    if (ended) return;
    ended = true;
    const statusCode = Number(statusCodeValue || 0);
    if (error || statusCode >= 500) {
      span.setStatus({
        code: SpanStatusCode.ERROR,
        ...(error instanceof Error ? { message: error.message } : {}),
      });
      if (error instanceof Error) span.recordException(error);
    } else {
      span.setStatus({ code: SpanStatusCode.OK });
    }
    if (statusCode) span.setAttribute('http.response.status_code', statusCode);
    span.end();
  };
  return { span, activeContext, end };
}

export async function withActiveContext(activeContext, fn) {
  return context.with(activeContext, fn);
}
